app.ts

```diff
@@ -0,0 +1,216 @@
+            }
+            getRecords(startIndex, startIndex + end);
+        }
+    };
+
+    nextbtn.onclick = function() {
+        let end = tableRecordCount - 1;
+        if (recordCount > startIndex + tableRecordCount) {
+            startIndex += tableRecordCount;
+            if (recordCount < startIndex + tableRecordCount) {
+                startIndex = recordCount - tableRecordCount;
+            }
+            getRecords(startIndex, startIndex + end);
+        }
+    };
+
+    prevbtn.onclick = function() {
+        if (startIndex >= tableRecordCount) {
+            startIndex -= tableRecordCount;
+        } else if (startIndex < tableRecordCount) {
+            startIndex = 0;
+        }
+        getRecords(startIndex, startIndex + tableRecordCount - 1);
+    };
+
+    $("#canvas").append(fromField,searchbtn, prevbtn, nextbtn);
+
+    initiate();
+
+    /**
+     * Initiates the HTTP requests to obtain the records and column values necessary for the table.
+     */
+    function initiate() {
+        // get number of records:
+        $.ajax({
+            url: "http://localhost:2050/recordCount",
+            success: function (result) {
+                recordCount = parseInt(JSON.parse(result), 10);
+                // get columns:
+                $.ajax({
+                    url: "http://localhost:2050/columns",
+                    success: function (result) {
+                        columns = JSON.parse(result);
+                        numColumns = getSize(columns);
+                        // get first page of records and display them:
+                        adjustTableRecordCount();
+                        getRecords(0, tableRecordCount - 1);
+                    },
+                    error: function (err) {
+                        $("body").text("Error: " + err.status + " " + err.statusText);
+                    }
+                });
+            },
+            error: function (err) {
+                $("body").text("Error: " + err.status + " " + err.statusText);
+            }
+        });
+    }
+
+    /**
+     * Generates a table populated with data, distrubuted evenly.
+     * @param data The data records retrieved from the server, structured as a 2D object array.
+     * @param columns An object array containing the column heading values.
+     */
+    function generateTable(data: any, columns: any) {
+        $("table").remove(); // Remove previous table
+        let dataSize = getSize(data);
+        let table = document.createElement("table");
+        let tbdy = document.createElement("tbody");
+        table.appendChild(generateHeadings(columns));
+        for (let i = 0; i < tableRecordCount; i++){
+            let tr = document.createElement("tr");
+            for (let j = 0; j < numColumns; j++){
+                let td = document.createElement("td");
+                if (dataSize > i)
+                    td.appendChild(document.createTextNode(data[i][j]));
```

Comment:
I have performance concerns about N^2 operations like this, especially because I know that there is some bottleneck with working with Javascript's native DOM API that I am not entirely familiar with - leading me to conclude this to perhaps be a slightly expensive call.
I know that in the historic IMQS codebase, we use a lot of string generation in order to minimize the number of raw appends that we do (i.e. generate the raw HTML in code and then only make 1 dom call), and that is how it was done in the past.
If you want to stick with the native JS API, I attached a link that speaks about documentfragments.
https://coderwall.com/p/o9ws2g/why-you-should-always-append-dom-elements-using-documentfragments